pytorch_lightning/core/lightning.py

```diff
@@ -1952,18 +1863,26 @@ def to_torchscript(
 
         return torchscript_module
 
-    @property
-    def hparams(self) -> Union[AttributeDict, dict, Namespace]:
-        if not hasattr(self, "_hparams"):
-            self._hparams = AttributeDict()
-        return self._hparams
+    def add_datamodule_hparams(self, datamodule: LightningDataModule):
+        """Add the hparams of a LightningDataModule to the hparams and hparams_initial of this module."""
```

Comment:
can you explain what this is and when this is used?
why can't the datamodule and lightning module maintain their own hparams?

please note the main value in `save_hyperparameters()` is that it captures the instantiation of the module, so we can do `LightningModule.load_from_checkpoint()`. 

but this does not exist for datamodules. 

in the assertions below I see

```python
    assert old_hparams == model.hparams
    assert old_hparams == model.hparams_initial
```

Isn't that problematic?

Reply:
The hyperparameters of the datamodule are added to the lightning module because only the hyperparameters of the lightning module are logged. The cited function is called inside the `attach_datamodule` function. This workflow is mainly inspired by the way I hacked the datamodule hyperparameters into lightning for my own projects.

Concerning the asserts: I had to add the hparams of the datamodule to `hparams_initial` because only the trainer only passes these to the logger. Considering that the added `hparams` are the initial hparams of the datamodule, I saw this as an acceptable solution. This problem would be solved by treating the hparams of the datamodule separately.

I could modify the PR to log the hparams of the lighting module and the datamodule separately. This would make it necessary to distinguish where the hparams came from in the logger. A simple solution would be adding a prefix when logging.

Reply:
Yes, I think the approach needs to be changed. It would be fine to merge the hparams at the logging stage but LightningModule should never contain hparams from a datamodule. Otherwise when we do

`LightningModule.load_from_checkpoint()` it would pass parameters to the init that don't exist. It means you would not be able to load your model and get a type error.

Reply:
Ok, I'll see if I find a way to do it. I hope it is okay to postpone making a '.load_from_checkpoint' function for datamodules, too.

Reply:
Thanks for your efforts. Yes that would not be required for this PR and could be discussed separately :)

So in summary, what we want here is:
- `model.save_hyperaparamters()` saves into `model.hparams`
- `datamodule.save_hyperparameters()` saves into `datamodule.hparams`
- no sharing of hyperparameters
- trainer grabs `model.hparams` and `datamodule.hparams` and logs the combined set

Reply:
Done. I adopted the behavior of the `add_datamodule_hparams` function that raised an error if there are duplicate hyperparameters.